### ISSUE TITLE
Incorrect sort order after a long uptime

### ISSUE THREAD

**Issue by dsigitov** [ClickHouse/ClickHouse#12748]
**Describe the bug**
After a long uptime and many INSERTS to a CollapsingMergeTree table, sort order sometimes become invalid for second expression in the ORDER BY clause.

For example,

```sql
SELECT attr_name_lang_ru, hex(MD5(attr_name_lang_ru)), hex(object_id), version, change_id, sign
FROM obj_ar_tirObject_history
ORDER BY attr_name_lang_ru, object_id, version, change_id
```

Record no | ok? | attr_name_lang_ru | hex(MD5(attr_name_lang_ru)) | hex(object_id) | version | change_id | sign
--- | --- | --- | --- | --- | --- | --- | ---  
1 | :heavy_check_mark: | Ульяновская область | 0FDD15815EA4F6C25DA9B4F0646523AA | 5F0733BABF4BB20007750252 | 6 | 0 | -1
2 | :heavy_check_mark: | Ульяновская область | 0FDD15815EA4F6C25DA9B4F0646523AA | 5F0733BABF4BB20007750252 | 6 | 0 | 1
3 | :heavy_check_mark: | Ульяновская область | 0FDD15815EA4F6C25DA9B4F0646523AA | 5F0733BABF4BB20007750252 | 7 | 0 | 1
4 | :x: | Хабаровский край | AF942E528062C1FF9AB72D85F78F0AE5 | 5F073385BF4BB200077500DE | 5 | 0 | 1
5 | :x: | Хабаровский край | AF942E528062C1FF9AB72D85F78F0AE5 | 5F073385BF4BB200077500DE | 6 | 0 | 1
6 | :x: | Хабаровский край | AF942E528062C1FF9AB72D85F78F0AE5 | 5F073385BF4BB200077500DE | 5 | 0 | -1
7 | :heavy_check_mark: | Ханты-Мансийский автономный округ | 667D8B75F146C6514EABE7AB6926FC57 | 5F073388BF4BB200077500F1 | 5 | 0 | -1
8 | :heavy_check_mark: | Ханты-Мансийский автономный округ | 667D8B75F146C6514EABE7AB6926FC57 | 5F073388BF4BB200077500F1 | 5 | 0 | 1
9 | :heavy_check_mark: | Ханты-Мансийский автономный округ | 667D8B75F146C6514EABE7AB6926FC57 | 5F073388BF4BB200077500F1 | 6 | 0 | 1

As you can see, the order of records 4, 5, 6 is incorrect: they have the same **attr_name_lang_ru** and **order_id**, but **version** is not ordered in increasing manner: record with **version** 6 is between two records with **version** 5.

I also included **hex(MD5(attr_name_lang_ru))** to ensure that strings in the **attr_name_lang_ru** column are binary equivalent.

**How to reproduce**
Unfortunately, I can't get a stable reproduction of this bug. It happens after a long uptime and many INSERTS to the table. I tried to copy the data to another table, but the order of records in the new table becomes correct. Also, OPTIMIZE FINAL fixes the issue.

* Current ClickHouse version: 20.5.3.27 in docker. We didn't encounter this bug in version 20.1.4.14.
* HTTP-interface
* All settings are default, except users

Minimal table structure:
```sql
create table obj_ar_tirObject_history
(
    object_id         FixedString(12),
    version           Int32,
    change_id         Int32,
    sign              Int8,
    attr_name_lang_ru String
) engine = CollapsingMergeTree(sign) ORDER BY (object_id, version, change_id);
```

**Expected behavior**
I expect record 5 to appear after records 4 and 6.

**Comment by den-crane** [comment]
try (for test and our understanding the cause of the issue) to query it with disabled features

set experimental_use_processors =0 
select ...

set optimize_read_in_order =0
select ...

**Comment by dsigitov** [comment]
> try (for test and our understanding the cause of the issue) to query it with disabled features

Unfortunately, none of these flags did help. I also tried all 8 possible combinations of them (0|1 for boolean flags and 1|8 for max_threads).

> can you also add _part column (it's a virtual column) to your select query and show the output?

Sure:

```sql
SELECT attr_name_lang_ru, hex(MD5(attr_name_lang_ru)), hex(object_id), version, change_id, sign, _part
FROM obj_ar_tirObject_history
ORDER BY attr_name_lang_ru, object_id, version, change_id
```
attr_name_lang_ru | hex(MD5(attr_name_lang_ru)) | hex(object_id) | version | change_id | sign | _part
--- | --- | --- | --- | --- | --- | ---
Ульяновская область | 0FDD15815EA4F6C25DA9B4F0646523AA | 5F0733BABF4BB20007750252 | 6 | 0 | -1 | all_51561_51603_7
Ульяновская область | 0FDD15815EA4F6C25DA9B4F0646523AA | 5F0733BABF4BB20007750252 | 6 | 0 | 1 | all_1_50110_191_50856
Ульяновская область | 0FDD15815EA4F6C25DA9B4F0646523AA | 5F0733BABF4BB20007750252 | 7 | 0 | 1 | all_51561_51603_7
Хабаровский край | AF942E528062C1FF9AB72D85F78F0AE5 | 5F073385BF4BB200077500DE | 5 | 0 | 1 | all_1_50110_191_50856
Хабаровский край | AF942E528062C1FF9AB72D85F78F0AE5 | 5F073385BF4BB200077500DE | 6 | 0 | 1 | all_51561_51603_7
Хабаровский край | AF942E528062C1FF9AB72D85F78F0AE5 | 5F073385BF4BB200077500DE | 5 | 0 | -1 | all_51561_51603_7
Ханты-Мансийский автономный округ | 667D8B75F146C6514EABE7AB6926FC57 | 5F073388BF4BB200077500F1 | 5 | 0 | -1 | all_51561_51603_7
Ханты-Мансийский автономный округ | 667D8B75F146C6514EABE7AB6926FC57 | 5F073388BF4BB200077500F1 | 5 | 0 | 1 | all_1_50110_191_50856
Ханты-Мансийский автономный округ | 667D8B75F146C6514EABE7AB6926FC57 | 5F073388BF4BB200077500F1 | 6 | 0 | 1 | all_51561_51603_7

**Comment by filimonov** [comment]
Do you insert both minus and plus row in a single insert?

**Comment by dsigitov** [comment]
> Do you insert both minus and plus row in a single insert?

Yes

> Do you insert in a single replica?

Yes, it's a standalone deployment.

**Comment by den-crane** [comment]
@dsigitov do you have reproducible example? Or maybe you can share you table data? backup copy of dataparts

**Comment by dsigitov** [comment]
@den-crane, yes. I prepared an excerpt from the original table. The issue is reproducable there.

[clickhouse-sorting-bug-reference.tar.gz](https://github.com/ClickHouse/ClickHouse/files/5001530/clickhouse-sorting-bug-reference.tar.gz)
[log.txt](https://github.com/ClickHouse/ClickHouse/files/5001551/log.txt)
